plugins/plugin-java/che-plugin-java-ext-lang-client/src/main/java/org/eclipse/che/ide/ext/java/client/progressor/ProgressMessagesHandler.java

```diff
@@ -0,0 +1,136 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.ide.ext.java.client.progressor;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static java.lang.System.currentTimeMillis;
+
+import com.google.gwt.core.client.Scheduler;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import com.google.web.bindery.event.shared.EventBus;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import org.eclipse.che.ide.api.workspace.event.WorkspaceStoppedEvent;
+import org.eclipse.che.ide.ext.java.client.progressor.background.BackgroundLoaderPresenter;
+import org.eclipse.che.ide.ext.java.shared.dto.progressor.ProgressReportDto;
+import org.eclipse.che.ide.util.Pair;
+
+/**
+ * Handler which receives messages from the jdt.ls server.
+ *
+ * @author Valeriy Svydenko
+ */
+@Singleton
+public class ProgressMessagesHandler {
+  private static final int DELAY_MS = 20_000; // wait before check if task was update
+  private final EventBus eventBus;
+  private final BackgroundLoaderPresenter progressResolver;
+  private Map<String, Pair<ProgressReportDto, Long>> progresses = new HashMap<>();
+
+  private ProgressReportDto currentProgress;
+
+  @Inject
+  public ProgressMessagesHandler(
+      EventBus eventBus,
+      ProgressorJsonRpcHandler progressorJsonRpcHandler,
+      BackgroundLoaderPresenter progressResolver) {
+    this.eventBus = eventBus;
+    this.progressResolver = progressResolver;
+
+    progressorJsonRpcHandler.addProgressReportHandler(this::handleProgressNotification);
+    handleOperations();
+  }
+
+  private void handleOperations() {
+    eventBus.addHandler(WorkspaceStoppedEvent.TYPE, event -> progressResolver.hide());
+  }
+
+  private void handleProgressNotification(ProgressReportDto progress) {
+    if (isNullOrEmpty(progress.getId())) {
+      return;
+    }
+    String taskId = progress.getId();
+
+    if (taskFinished(progress)) {
+      progresses.remove(taskId);
+      progressResolver.updateProgressBar(progress);
+      progressResolver.removeProgress(progress);
+      updateCurrentTask();
+      return;
+    }
+
+    if (progress.isComplete()) {
+      return;
+    }
+
+    if (progresses.isEmpty()) {
+      addFirstProgress(progress);
+    }
+
+    progresses.put(taskId, Pair.of(progress, currentTimeMillis()));
+    progressResolver.addProgress(progress);
+
+    String currentTaskId = currentProgress.getId();
+
+    if (taskId.equals(currentTaskId)) {
+      currentProgress = progress;
+      progressResolver.updateProgressBar(currentProgress);
+    }
+    progressResolver.show();
+  }
+
+  private void startProcessesCleaner() {
+    Scheduler.get()
+        .scheduleFixedDelay(
+            () -> {
+              for (Iterator<Map.Entry<String, Pair<ProgressReportDto, Long>>> it =
```

Comment:
I am a bit worried about accessing `progresses` from different threads.

Reply:
I think here is a single thread.

Reply:
The main and the scheduled threads

Reply:
this compiled JS is single-threaded, no?

Reply:
@vparfonov What do you think?

Reply:
@tolusha  Yes JS working in single thread, so no sweat :)